### PR TITLE
Add support for congratulation letters for non-fa programs

### PR DESCRIPTION
**Pull request by tasawernawaz**
#### What are the relevant tickets?
#4247 

#### What's this PR do?
Add support for congratulation letter for non-fa programs

#### Any background context you want to provide?
Read RFC at #4258 

**How to test:**

- In the CMS admin create `ProgramCertificateSignatories`, `program_letter_text` and `program_letter_logo` for the program.
- Set `program.financial_aid_availability = False`
- Enroll user in program.
- Create a passed course with a grade by running `atler_data` command with `set_past_run_to_passed` for all courses in enrolled program.
- Create course certificate for all courses:
- `MicromastersCourseCertificate.objects.create(user=<user>, course=<course>)`
- Congratulations letter should automatically created after having certificate in all courses.

#### Screenshots (if appropriate)

This is how it will looks like.

**Dashboard:**
screen shot 2019-03-04 at 5 59 24 pm

**Congratulations Letter:**
screen shot 2019-02-28 at 4 18 08 pm

### CODE REVIEW

**Comment by tasawernawaz** [comment]
@pdpinch We have a model `ProgramCertificateSignatories`, will there be same signatories for letters ? Otherwise we can create separate model for letter signatories.

**Comment by pdpinch** [comment]
Separate signatories.

**Comment by tasawernawaz** [comment]
@pdpinch Please confirm, Do non-fa program users have certificates ?

also, I need MIT logo image shown in the used in attached pdf. Assuming that logos will remain same for all letters.

**Comment by pdpinch** [comment]
Non-fa learners get course and program certificates on EdX.org. We don’t crest either for them. We track their course certificates (but not program certificates) in a EdX cache table. 

I’ll get you the logo, but it’s custom per letter (custom colors that match the program logo).

**Comment by tasawernawaz** [comment]
@pdpinch If the logo is custom per program then I guess we have to save it in CMS through ProgramPage? right ?

**Comment by pdpinch** [comment]
That’s what I had in mind, similar to the signatures. . 

> @pdpinch <https://github.com/pdpinch> If the logo is custom per program then I guess we have to save it in CMS through ProgramPage? right ?

**Comment by tasawernawaz** [comment]
@pdpinch This is ready for review, Could you please ask someone to take a look.

P.S: Meanwhile I can add some more tests to increase coverage.

File: grades/api.py [hunk 1]
```diff
@@ -248,6 +248,46 @@ def generate_program_certificate(user, program):
...
+    for each course in the program and program is non-fa.
+
+    Args:
+        user (User): a Django user.
+        program (programs.models.Program): program where the user is enrolled.
+    """
+
+    if program.financial_aid_availability:
+        log.info(
+            'Congratulation letter can not be created for [%s], as given program [%s] if not non-fa',
```
Comment (gsidebo) [inline grades/api.py]: `'Congratulation letter is only available for non-financial aid programs'`. Also, I think this should be a `log.error`

File: grades/api.py [hunk 2]
```diff
@@ -248,6 +248,46 @@ def generate_program_certificate(user, program):
...
+            user.username,
+            program.title
+        )
+        return
+
+    if MicromastersProgramCommendation.objects.filter(user=user, program=program).exists():
+        log.error('User [%s] already has a letter for program [%s]', user, program)
+        return
+
+    courses_in_program_ids = set(program.course_set.all().values_list('id', flat=True))
```
Comment (gsidebo) [inline grades/api.py]: Rename to `program_course_ids`

File: grades/management/commands/generate_program_letter.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+from courses.models import Program
+from dashboard.models import ProgramEnrollment
+from grades.api import generate_program_letter
+
+
+class Command(BaseCommand):
+    """
+    Finds all completed non-FA programs
+    """
+    help = "Finds all users that completed the program"
```
Comment (gsidebo) [inline grades/management/commands/generate_program_letter.py]: Both the docstring and help text should match the module docstring

File: grades/management/commands/generate_program_letter.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    def handle(self, *args, **kwargs):  # pylint: disable=unused-argument
+        programs = Program.objects.filter(live=True, financial_aid_availability=False)
+        for program in programs:
+            if program.has_frozen_grades_for_all_courses():
+                enrollments = ProgramEnrollment.objects.filter(program=program)
```
Comment (gsidebo) [inline grades/management/commands/generate_program_letter.py]: Please add `select_related('user')` to the query

File: grades/signals.py [hunk 5]
```diff
@@ -20,6 +20,17 @@ def handle_create_coursecertificate(sender, instance, created, **kwargs):  # pyl
...
 
+@receiver(post_save, sender=MicromastersCourseCertificate, dispatch_uid="coursecertificate_post_save_for_letters")
+def handle_for_createprogram_letters(sender, instance, created, **kwargs):  # pylint: disable=unused-argument
+    """
+    When a MicromastersCourseCertificate model is created
+    """
+    if created:
+        user = instance.user
+        program = instance.course.program
+        transaction.on_commit(lambda: generate_program_letter(user, program))
```
Comment (gsidebo) [inline grades/signals.py]: This should only be run if the program is non-financial aid

File: static/scss/letter.scss [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  .signatories {
+    .signatory {
+      margin-right: 15px;
+      display: inline-block;
+    }
+
+    .sig-image {
+      height: 100px;
```
Comment (gsidebo) [inline static/scss/letter.scss]: Is this necessary since you're already adjusting the size of the image in the template with the `image` templatetag?

Comment (tasawernawaz) [reply static/scss/letter.scss]: Just to make sure that even if images have different sizes, they should render inline with each other.

**Comment by tasawernawaz** [comment]
@gsidebo Feedback addressed, and few more tests added. Please take another look.

**Comment by tasawernawaz** [comment]
@gsidebo You are right, I have discussed this with @pdpinch on call. We were supposed to do this in future. As initially we were not sure that each program is going to have different letter text or not. I think at this point if it's clear, I can make this change.

**Comment by gsidebo** [comment]
Functionality looks great. Reviewing code now

File: grades/management/commands/generate_program_letter.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+class Command(BaseCommand):
+    """
+    Finds all users, those have completed non-FA programs and generate letter for them
+    """
+    help = "Finds all users those have completed the non-FA program and generate letter for them."
+
+    def handle(self, *args, **kwargs):  # pylint: disable=unused-argument
+        programs = Program.objects.filter(live=True, financial_aid_availability=False)
+        for program in programs:
+            if program.has_frozen_grades_for_all_courses():
```
[the remark this answers is not in the document]
Comment (tasawernawaz) [reply grades/management/commands/generate_program_letter.py]: @gsidebo changes implemented.

**Pull request merged**